File: typescript/src/networks/assets.ts
/** Canonical payment-asset catalog for supported BNB Chain networks. */

import { getAddress as toChecksumAddress } from "viem";
import {
  BSC_MAINNET_CHAIN_ID,
  BSC_TESTNET_CHAIN_ID,
  PAYMENT_TOKEN_EIP712_NAME,
  PAYMENT_TOKEN_EIP712_VERSION,
} from "./addresses.js";

export const AssetId = Object.freeze({
  U: "U",
  TEST_U: "TEST_U",
  BINANCE_PEG_USDC: "BINANCE_PEG_USDC",
  BINANCE_PEG_USDT: "BINANCE_PEG_USDT",
  TEST_USDC: "TEST_USDC",
  TEST_USDT: "TEST_USDT",
} as const);

export type AssetId = (typeof AssetId)[keyof typeof AssetId];
export type AssetAlias = "U" | "USDC" | "USDT";
export type B402TransferMethod = "eip3009" | "permit2-exact";

export interface EIP3009Domain {
  readonly name: string;
  readonly version: string;
}

export interface B402Kind {
  readonly method: B402TransferMethod;
  readonly name: string;
  readonly version: string;
}

export interface PaymentAsset {
  readonly chainId: number;
  readonly assetId: AssetId;
  readonly symbol: string;
  readonly address: `0x${string}`;
  readonly decimals: number;
  readonly b402Methods: readonly B402TransferMethod[];
  /**
   * Optional for source compatibility with pre-kind object literals.
   * AssetCatalog rejects omission and returns CatalogPaymentAsset instead.
   */
  readonly b402Kinds?: readonly B402Kind[];
  readonly eip3009Domain: EIP3009Domain | null;
  readonly isDefault: boolean;
}

/** A catalog-validated asset with complete B402 kind identities. */
export interface CatalogPaymentAsset extends PaymentAsset {
  readonly b402Kinds: readonly B402Kind[];
}

const CANONICAL_IDS = new Set<string>(Object.values(AssetId));

export function parseAssetId(value: string): AssetId {
  if (!CANONICAL_IDS.has(value)) {
    throw new Error(`unknown canonical AssetId: ${JSON.stringify(value)}`);
  }
  return value as AssetId;
}

/** Fail-closed lookup index keyed by network, canonical id, and address. */
export class AssetCatalog {
  readonly #byKey = new Map<string, CatalogPaymentAsset>();
  readonly #byAddress = new Map<string, CatalogPaymentAsset>();
  readonly #byChain = new Map<number, readonly CatalogPaymentAsset[]>();

  constructor(assets: readonly PaymentAsset[]) {
    const mutableByChain = new Map<number, CatalogPaymentAsset[]>();

    for (const input of assets) {
      const assetId = parseAssetId(input.assetId);
      const key = `${input.chainId}:${assetId}`;
      if (this.#byKey.has(key)) {
        throw new Error(
          `duplicate catalog key: chain_id=${input.chainId}, asset_id=${assetId}`,
        );
      }

      const address = toChecksumAddress(input.address);
      if (address !== input.address) {
        throw new Error(`catalog address is not checksummed: ${input.address}`);
      }

      const b402Kinds = validateB402Metadata(input);
      const addressKey = `${input.chainId}:${address.toLowerCase()}`;
      if (this.#byAddress.has(addressKey)) {
        throw new Error(
          `duplicate catalog address: chain_id=${input.chainId}, address=${address}`,
        );
      }

      const asset = Object.freeze({
        ...input,
        assetId,
        address,
        b402Methods: Object.freeze([...input.b402Methods]),
        b402Kinds: Object.freeze(
          b402Kinds.map((kind) => Object.freeze({ ...kind })),
        ),
        eip3009Domain:
          input.eip3009Domain === null
            ? null
            : Object.freeze({ ...input.eip3009Domain }),
      });
      this.#byKey.set(key, asset);
      this.#byAddress.set(addressKey, asset);
      const chainAssets = mutableByChain.get(asset.chainId) ?? [];
      chainAssets.push(asset);
      mutableByChain.set(asset.chainId, chainAssets);
    }

    for (const [chainId, chainAssets] of mutableByChain) {
      this.#byChain.set(chainId, Object.freeze([...chainAssets]));
    }
    Object.freeze(this);
  }

  #requireChain(chainId: number): void {
    if (!this.#byChain.has(chainId)) {
      throw new Error(`no asset catalog registered for chain_id=${chainId}`);
    }
  }

  get(chainId: number, assetId: AssetId | string): CatalogPaymentAsset {
    this.#requireChain(chainId);
    const canonical = parseAssetId(assetId);
    const asset = this.#byKey.get(`${chainId}:${canonical}`);
    if (asset === undefined) {
      throw new Error(
        `AssetId ${canonical} is not available on chain_id=${chainId}`,
      );
    }
    return asset;
  }

  byAddress(chainId: number, inputAddress: string): CatalogPaymentAsset {
    this.#requireChain(chainId);
    let address: `0x${string}`;
    try {
      address = toChecksumAddress(inputAddress);
    } catch (error) {
      throw new Error(
        `asset address ${JSON.stringify(inputAddress)} is not registered on chain_id=${chainId}`,
        { cause: error },
      );
    }
    const asset = this.#byAddress.get(`${chainId}:${address.toLowerCase()}`);
    if (asset === undefined) {
      throw new Error(
        `asset address ${JSON.stringify(inputAddress)} is not registered on chain_id=${chainId}`,
      );
    }
    return asset;
  }

  list(chainId: number): readonly CatalogPaymentAsset[] {
    this.#requireChain(chainId);
    const assets = this.#byChain.get(chainId);
    if (assets === undefined) {
      throw new Error(`no asset catalog registered for chain_id=${chainId}`);
    }
    return assets;
  }
}

const B402_TRANSFER_METHODS = new Set<B402TransferMethod>([
  "eip3009",
  "permit2-exact",
]);

function validateB402Metadata(input: PaymentAsset): readonly B402Kind[] {
  const methods = input.b402Methods;
  if (new Set(methods).size !== methods.length) {
    throw new Error("duplicate B402 method in asset catalog");
  }
  if (methods.some((method) => !B402_TRANSFER_METHODS.has(method))) {
    throw new Error("unknown B402 method in asset catalog");
  }

  const kinds = input.b402Kinds ?? [];
  const kindMethods = kinds.map((kind) => kind.method);
  if (new Set(kindMethods).size !== kindMethods.length) {
    throw new Error("duplicate B402 kind identity in asset catalog");
  }
  if (kindMethods.some((method) => !B402_TRANSFER_METHODS.has(method))) {
    throw new Error("unknown B402 kind method in asset catalog");
  }
  if (
    kinds.some((kind) => kind.name.length === 0 || kind.version.length === 0)
  ) {
    throw new Error("B402 kind name and version must be non-empty");
  }

  const missing = methods.find((method) => !kindMethods.includes(method));
  if (missing !== undefined) {
    throw new Error(`missing B402 kind identity for method=${missing}`);
  }
  const extra = kindMethods.find((method) => !methods.includes(method));
  if (extra !== undefined) {
    throw new Error(`extra B402 kind identity for method=${extra}`);
  }

  const eip3009Kind = kinds.find((kind) => kind.method === "eip3009");
  if (eip3009Kind === undefined) {
    if (input.eip3009Domain !== null) {
      throw new Error("EIP-3009 domain requires an eip3009 B402 method");
    }
  } else if (
    input.eip3009Domain === null ||
    eip3009Kind.name !== input.eip3009Domain.name ||
    eip3009Kind.version !== input.eip3009Domain.version
  ) {
    throw new Error("EIP-3009 kind must match eip3009Domain exactly");
  }
  return kinds;
}

const EIP3009_DOMAIN = Object.freeze({
  name: PAYMENT_TOKEN_EIP712_NAME,
  version: PAYMENT_TOKEN_EIP712_VERSION,
});

export const ASSET_CATALOG = new AssetCatalog([
  {
    chainId: BSC_MAINNET_CHAIN_ID,
    assetId: AssetId.U,
    symbol: "U",
    address: "0xcE24439F2D9C6a2289F741120FE202248B666666",
    decimals: 18,
    b402Methods: ["eip3009", "permit2-exact"],
    b402Kinds: [
      { method: "eip3009", name: "United Stables", version: "1" },
      { method: "permit2-exact", name: "United Stables", version: "1" },
    ],
    eip3009Domain: EIP3009_DOMAIN,
    isDefault: true,
  },
  {
    chainId: BSC_MAINNET_CHAIN_ID,
    assetId: AssetId.BINANCE_PEG_USDC,
    symbol: "USDC",
    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
    decimals: 18,
    b402Methods: ["permit2-exact"],
    b402Kinds: [{ method: "permit2-exact", name: "USD Coin", version: "1" }],
    eip3009Domain: null,
    isDefault: false,
  },
  {
    chainId: BSC_MAINNET_CHAIN_ID,
    assetId: AssetId.BINANCE_PEG_USDT,
    symbol: "USDT",
    address: "0x55d398326f99059fF775485246999027B3197955",
    decimals: 18,
    b402Methods: ["permit2-exact"],
    b402Kinds: [{ method: "permit2-exact", name: "Tether USD", version: "1" }],
    eip3009Domain: null,
    isDefault: false,
  },
  {
    chainId: BSC_TESTNET_CHAIN_ID,
    assetId: AssetId.TEST_U,
    symbol: "U",
    address: "0xc70B8741B8B07A6d61E54fd4B20f22Fa648E5565",
    decimals: 18,
    b402Methods: ["eip3009"],
    b402Kinds: [{ method: "eip3009", name: "United Stables", version: "1" }],
    eip3009Domain: EIP3009_DOMAIN,
    isDefault: true,
  },
  {
    chainId: BSC_TESTNET_CHAIN_ID,
    assetId: AssetId.TEST_USDC,
    symbol: "USDC",
    address: "0xEC1C60D64a06896Df296438c12edD14E974FDE47",
    decimals: 6,
    b402Methods: ["permit2-exact"],
    b402Kinds: [{ method: "permit2-exact", name: "USD Coin", version: "1" }],
    eip3009Domain: null,
    isDefault: false,
  },
  {
    chainId: BSC_TESTNET_CHAIN_ID,
    assetId: AssetId.TEST_USDT,
    symbol: "USDT",
    address: "0x337610d27c682E347C9cD60BD4b3b107C9d34dDd",
    decimals: 18,
    b402Methods: ["permit2-exact"],
    b402Kinds: [{ method: "permit2-exact", name: "USDT Token", version: "1" }],
    eip3009Domain: null,
    isDefault: false,
  },
]);

const ALIASES: Readonly<Record<number, Readonly<Record<AssetAlias, AssetId>>>> =
  Object.freeze({
    [BSC_MAINNET_CHAIN_ID]: Object.freeze({
      U: AssetId.U,
      USDC: AssetId.BINANCE_PEG_USDC,
      USDT: AssetId.BINANCE_PEG_USDT,
    }),
    [BSC_TESTNET_CHAIN_ID]: Object.freeze({
      U: AssetId.TEST_U,
      USDC: AssetId.TEST_USDC,
      USDT: AssetId.TEST_USDT,
    }),
  });

export function resolveAssetAlias(chainId: number, alias: string): AssetId {
  const aliases = ALIASES[chainId];
  if (aliases === undefined) {
    throw new Error(`no asset catalog registered for chain_id=${chainId}`);
  }
  const resolved = aliases[alias as AssetAlias];
  if (resolved === undefined) {
    throw new Error(
      `unknown asset alias ${JSON.stringify(alias)} for chain_id=${chainId}`,
    );
  }
  return resolved;
}

export function getAsset(
  chainId: number,
  assetId: AssetId | string,
): CatalogPaymentAsset {
  return ASSET_CATALOG.get(chainId, assetId);
}

export function getAssetByAddress(
  chainId: number,
  address: string,
): CatalogPaymentAsset {
  return ASSET_CATALOG.byAddress(chainId, address);
}

export function listAssets(chainId: number): readonly CatalogPaymentAsset[] {
  return ASSET_CATALOG.list(chainId);
}

const DECIMAL_AMOUNT = /^[0-9]+(?:\.[0-9]+)?$/;

export function toAssetAtomic(
  chainId: number,
  assetId: AssetId | string,
  amount: string,
): bigint {
  if (typeof amount !== "string") {
    throw new TypeError("asset amount must be a decimal string");
  }
  if (!DECIMAL_AMOUNT.test(amount)) {
    throw new Error(`invalid decimal amount: ${JSON.stringify(amount)}`);
  }

  const asset = getAsset(chainId, assetId);
  const [whole, fraction = ""] = amount.split(".");
  if (fraction.length > asset.decimals) {
    throw new Error(
      `asset amount exceeds ${asset.decimals} decimal places: ${JSON.stringify(amount)}`,
    );
  }
  return (
    BigInt(whole) * 10n ** BigInt(asset.decimals) +
    BigInt(fraction.padEnd(asset.decimals, "0") || "0")
  );
}

File: python/bnbagent/networks/assets.py
"""Canonical payment-asset catalog for supported BNB Chain networks."""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, replace
from enum import Enum
from types import MappingProxyType
from typing import Literal

from web3 import Web3

from .addresses import (
    BSC_MAINNET_CHAIN_ID,
    BSC_TESTNET_CHAIN_ID,
    PAYMENT_TOKEN_EIP712_NAME,
    PAYMENT_TOKEN_EIP712_VERSION,
)

B402TransferMethod = Literal["eip3009", "permit2-exact"]
AssetAlias = Literal["U", "USD1", "USDC", "USDT"]
PaymentAssetAvailability = Literal["active", "placeholder"]


class AssetId(str, Enum):
    """Stable asset identities used by SDK APIs and persisted configuration."""

    U = "U"
    TEST_U = "TEST_U"
    USD1 = "USD1"
    BINANCE_PEG_USDC = "BINANCE_PEG_USDC"
    BINANCE_PEG_USDT = "BINANCE_PEG_USDT"
    TEST_USDC = "TEST_USDC"
    TEST_USDT = "TEST_USDT"


@dataclass(frozen=True)
class EIP3009Domain:
    """Verified EIP-712 domain metadata for an EIP-3009 token."""

    name: str
    version: str


@dataclass(frozen=True)
class B402Kind:
    """Exact B402 scheme identity for one transfer method."""

    method: B402TransferMethod
    name: str
    version: str


class PaymentAssetUnavailableError(KeyError):
    """Raised when metadata names an asset unavailable for payment queries."""

    def __init__(self, chain_id: int, asset_id: AssetId) -> None:
        super().__init__(f"AssetId {asset_id.value} is unavailable on chain_id={chain_id}")
        self.chain_id = chain_id
        self.asset_id = asset_id


@dataclass(frozen=True)
class PaymentAsset:
    """Immutable metadata for one canonical asset on one network."""

    chain_id: int
    asset_id: AssetId
    symbol: str
    address: str
    decimals: int
    b402_methods: tuple[B402TransferMethod, ...]
    eip3009_domain: EIP3009Domain | None
    is_default: bool
    # A trailing default preserves direct construction compatibility. An
    # AssetCatalog still rejects missing identities for declared methods.
    b402_kinds: tuple[B402Kind, ...] = ()
    # A trailing default preserves legacy object construction semantics.
    availability: PaymentAssetAvailability = "active"
    # B402 execution facts default to the ERC-8183 address and decimals.
    b402_address: str | None = None
    b402_decimals: int | None = None


B402PaymentAsset = PaymentAsset


class AssetCatalog:
    """Fail-closed lookup index keyed by network, canonical id, and address."""

    def __init__(self, assets: Sequence[PaymentAsset]) -> None:
        metadata_by_key: dict[tuple[int, AssetId], PaymentAsset] = {}
        metadata_by_address: dict[tuple[int, str], PaymentAsset] = {}
        active_by_key: dict[tuple[int, AssetId], PaymentAsset] = {}
        active_by_address: dict[tuple[int, str], PaymentAsset] = {}
        active_b402_by_key: dict[tuple[int, AssetId], B402PaymentAsset] = {}
        active_b402_by_address: dict[tuple[int, str], B402PaymentAsset] = {}
        active_by_chain: dict[int, list[PaymentAsset]] = {}

        for asset in assets:
            if not isinstance(asset.asset_id, AssetId):
                raise ValueError(f"asset_id must be a canonical AssetId: {asset.asset_id!r}")
            key = (asset.chain_id, asset.asset_id)
            if key in metadata_by_key:
                raise ValueError(
                    f"duplicate catalog key: chain_id={asset.chain_id}, "
                    f"asset_id={asset.asset_id.value}"
                )

            if not Web3.is_checksum_address(asset.address):
                raise ValueError(f"catalog address is not checksummed: {asset.address}")

            self._validate_availability(asset)
            self._validate_b402_metadata(asset)
            b402_address = asset.b402_address or asset.address
            b402_decimals = asset.decimals if asset.b402_decimals is None else asset.b402_decimals
            if not Web3.is_checksum_address(b402_address):
                raise ValueError(f"catalog B402 address is not checksummed: {b402_address}")
            if isinstance(b402_decimals, bool) or not isinstance(b402_decimals, int):
                raise ValueError("catalog B402 decimals must be a non-negative integer")
            if b402_decimals < 0:
                raise ValueError("catalog B402 decimals must be a non-negative integer")
            address_key = (asset.chain_id, asset.address.lower())
            if address_key in metadata_by_address:
                raise ValueError(
                    f"duplicate catalog address: chain_id={asset.chain_id}, "
                    f"address={asset.address}"
                )

            metadata_by_key[key] = asset
            metadata_by_address[address_key] = asset
            active_by_chain.setdefault(asset.chain_id, [])
            if asset.availability == "active":
                if b402_address.lower() == _ZERO_ADDRESS:
                    raise ValueError("active B402 asset cannot use zero address")
                b402_address_key = (asset.chain_id, b402_address.lower())
                if b402_address_key in active_b402_by_address:
                    raise ValueError(
                        f"duplicate catalog B402 address: chain_id={asset.chain_id}, "
                        f"address={b402_address}"
                    )
                b402_asset = replace(
                    asset,
                    address=b402_address,
                    decimals=b402_decimals,
                    b402_address=b402_address,
                    b402_decimals=b402_decimals,
                )
                active_by_key[key] = asset
                active_by_address[address_key] = asset
                active_b402_by_key[key] = b402_asset
                active_b402_by_address[b402_address_key] = b402_asset
                active_by_chain[asset.chain_id].append(asset)

        self._metadata_by_key: Mapping[tuple[int, AssetId], PaymentAsset] = MappingProxyType(
            metadata_by_key
        )
        self._metadata_by_address: Mapping[tuple[int, str], PaymentAsset] = MappingProxyType(
            metadata_by_address
        )
        self._active_by_key: Mapping[tuple[int, AssetId], PaymentAsset] = MappingProxyType(
            active_by_key
        )
        self._active_by_address: Mapping[tuple[int, str], PaymentAsset] = MappingProxyType(
            active_by_address
        )
        self._active_b402_by_key: Mapping[tuple[int, AssetId], B402PaymentAsset] = (
            MappingProxyType(active_b402_by_key)
        )
        self._active_b402_by_address: Mapping[tuple[int, str], B402PaymentAsset] = (
            MappingProxyType(active_b402_by_address)
        )
        self._active_by_chain: Mapping[int, tuple[PaymentAsset, ...]] = MappingProxyType(
            {chain_id: tuple(chain_assets) for chain_id, chain_assets in active_by_chain.items()}
        )

    @staticmethod
    def _validate_availability(asset: PaymentAsset) -> None:
        if asset.availability not in ("active", "placeholder"):
            raise ValueError("unknown payment asset availability")
        if asset.availability == "active":
            if asset.address.lower() == _ZERO_ADDRESS:
                raise ValueError("active asset cannot use zero address")
            return
        if asset.address.lower() != _ZERO_ADDRESS:
            raise ValueError("placeholder asset must use zero address")
        if asset.b402_methods or asset.b402_kinds:
            raise ValueError("placeholder asset cannot declare B402 methods")
        if asset.eip3009_domain is not None:
            raise ValueError("placeholder asset cannot declare an EIP-3009 domain")
        if asset.is_default:
            raise ValueError("placeholder asset cannot be the default asset")

    @staticmethod
    def _validate_b402_metadata(asset: PaymentAsset) -> None:
        supported_methods = frozenset(("eip3009", "permit2-exact"))
        methods = asset.b402_methods
        if not isinstance(methods, tuple):
            raise ValueError("b402_methods must be an immutable tuple")
        if len(set(methods)) != len(methods):
            raise ValueError("duplicate B402 method in asset catalog")
        if any(method not in supported_methods for method in methods):
            raise ValueError("unknown B402 method in asset catalog")

        kinds = asset.b402_kinds
        if not isinstance(kinds, tuple) or any(not isinstance(kind, B402Kind) for kind in kinds):
            raise ValueError("b402_kinds must be an immutable tuple of B402Kind")
        kind_methods = tuple(kind.method for kind in kinds)
        if len(set(kind_methods)) != len(kind_methods):
            raise ValueError("duplicate B402 kind identity in asset catalog")
        if any(kind.method not in supported_methods for kind in kinds):
            raise ValueError("unknown B402 kind method in asset catalog")
        if any(not kind.name or not kind.version for kind in kinds):
            raise ValueError("B402 kind name and version must be non-empty")

        missing = set(methods).difference(kind_methods)
        if missing:
            raise ValueError(f"missing B402 kind identity for method={sorted(missing)[0]}")
        extra = set(kind_methods).difference(methods)
        if extra:
            raise ValueError(f"extra B402 kind identity for method={sorted(extra)[0]}")

        eip3009_kind = next((kind for kind in kinds if kind.method == "eip3009"), None)
        if eip3009_kind is None:
            if asset.eip3009_domain is not None:
                raise ValueError("EIP-3009 domain requires an eip3009 B402 method")
        elif asset.eip3009_domain is None or (
            eip3009_kind.name,
            eip3009_kind.version,
        ) != (
            asset.eip3009_domain.name,
            asset.eip3009_domain.version,
        ):
            raise ValueError("EIP-3009 kind must match eip3009_domain exactly")

    def _require_chain(self, chain_id: int) -> None:
        if chain_id not in self._active_by_chain:
            raise KeyError(f"no asset catalog registered for chain_id={chain_id}")

    def get(self, chain_id: int, asset_id: AssetId | str) -> PaymentAsset:
        asset = self.get_metadata(chain_id, asset_id)
        if asset.availability == "placeholder":
            raise PaymentAssetUnavailableError(chain_id, asset.asset_id)
        return self._active_by_key[(chain_id, asset.asset_id)]

    def get_metadata(self, chain_id: int, asset_id: AssetId | str) -> PaymentAsset:
        self._require_chain(chain_id)
        canonical = parse_asset_id(asset_id)
        try:
            return self._metadata_by_key[(chain_id, canonical)]
        except KeyError as exc:
            raise KeyError(
                f"AssetId {canonical.value} is not available on chain_id={chain_id}"
            ) from exc

    def by_address(self, chain_id: int, address: str) -> PaymentAsset:
        self._require_chain(chain_id)
        try:
            address_key = Web3.to_checksum_address(address).lower()
        except (TypeError, ValueError) as exc:
            raise KeyError(
                f"asset address {address!r} is not registered on chain_id={chain_id}"
            ) from exc
        try:
            return self._active_by_address[(chain_id, address_key)]
        except KeyError as exc:
            raise KeyError(
                f"asset address {address!r} is not registered on chain_id={chain_id}"
            ) from exc

    def get_b402(self, chain_id: int, asset_id: AssetId | str) -> B402PaymentAsset:
        asset = self.get(chain_id, asset_id)
        return self._active_b402_by_key[(chain_id, asset.asset_id)]

    def by_b402_address(self, chain_id: int, address: str) -> B402PaymentAsset:
        self._require_chain(chain_id)
        try:
            address_key = Web3.to_checksum_address(address).lower()
        except (TypeError, ValueError) as exc:
            raise KeyError(
                f"B402 asset address {address!r} is not registered on chain_id={chain_id}"
            ) from exc
        try:
            return self._active_b402_by_address[(chain_id, address_key)]
        except KeyError as exc:
            raise KeyError(
                f"B402 asset address {address!r} is not registered on chain_id={chain_id}"
            ) from exc

    def list(self, chain_id: int) -> tuple[PaymentAsset, ...]:
        self._require_chain(chain_id)
        return self._active_by_chain[chain_id]


_DOMAIN = EIP3009Domain(
    name=PAYMENT_TOKEN_EIP712_NAME,
    version=PAYMENT_TOKEN_EIP712_VERSION,
)

_ZERO_ADDRESS = "0x0000000000000000000000000000000000000000"

ASSET_CATALOG = AssetCatalog(
    (
        PaymentAsset(
            chain_id=BSC_MAINNET_CHAIN_ID,
            asset_id=AssetId.U,
            symbol="U",
            address="0xcE24439F2D9C6a2289F741120FE202248B666666",
            decimals=18,
            b402_methods=("eip3009", "permit2-exact"),
            eip3009_domain=_DOMAIN,
            is_default=True,
            b402_kinds=(
                B402Kind("eip3009", "United Stables", "1"),
                B402Kind("permit2-exact", "United Stables", "1"),
            ),
        ),
        PaymentAsset(
            chain_id=BSC_MAINNET_CHAIN_ID,
            asset_id=AssetId.USD1,
            symbol="USD1",
            address="0x8d0D000Ee44948FC98c9B98A4FA4921476f08B0d",
            decimals=18,
            availability="active",
            b402_methods=("eip3009",),
            eip3009_domain=EIP3009Domain("World Liberty Financial USD", "1"),
            is_default=False,
            b402_kinds=(B402Kind("eip3009", "World Liberty Financial USD", "1"),),
        ),
        PaymentAsset(
            chain_id=BSC_MAINNET_CHAIN_ID,
            asset_id=AssetId.BINANCE_PEG_USDC,
            symbol="USDC",
            address="0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
            decimals=18,
            b402_methods=("permit2-exact",),
            eip3009_domain=None,
            is_default=False,
            b402_kinds=(B402Kind("permit2-exact", "USD Coin", "1"),),
        ),
        PaymentAsset(
            chain_id=BSC_MAINNET_CHAIN_ID,
            asset_id=AssetId.BINANCE_PEG_USDT,
            symbol="USDT",
            address="0x55d398326f99059fF775485246999027B3197955",
            decimals=18,
            b402_methods=("permit2-exact",),
            eip3009_domain=None,
            is_default=False,
            b402_kinds=(B402Kind("permit2-exact", "Tether USD", "1"),),
        ),
        PaymentAsset(
            chain_id=BSC_TESTNET_CHAIN_ID,
            asset_id=AssetId.TEST_U,
            symbol="U",
            address="0xc70B8741B8B07A6d61E54fd4B20f22Fa648E5565",
            decimals=18,
            b402_address="0x330949Aed7d00FCe0558C64ED6FeC9792616cC39",
            b402_decimals=6,
            b402_methods=("eip3009",),
            eip3009_domain=_DOMAIN,
            is_default=True,
            b402_kinds=(B402Kind("eip3009", "United Stables", "1"),),
        ),
        PaymentAsset(
            chain_id=BSC_TESTNET_CHAIN_ID,
            asset_id=AssetId.TEST_USDC,
            symbol="USDC",
            address="0xEC1C60D64a06896Df296438c12edD14E974FDE47",
            decimals=6,
            b402_methods=("permit2-exact",),
            eip3009_domain=None,
            is_default=False,
            b402_kinds=(B402Kind("permit2-exact", "USD Coin", "1"),),
        ),
        PaymentAsset(
            chain_id=BSC_TESTNET_CHAIN_ID,
            asset_id=AssetId.TEST_USDT,
            symbol="USDT",
            address="0x337610d27c682E347C9cD60BD4b3b107C9d34dDd",
            decimals=18,
            b402_methods=("permit2-exact",),
            eip3009_domain=None,
            is_default=False,
            b402_kinds=(B402Kind("permit2-exact", "USDT Token", "1"),),
        ),
    )
)

_ALIASES: Mapping[int, Mapping[str, AssetId]] = MappingProxyType(
    {
        BSC_MAINNET_CHAIN_ID: MappingProxyType(
            {
                "U": AssetId.U,
                "USD1": AssetId.USD1,
                "USDC": AssetId.BINANCE_PEG_USDC,
                "USDT": AssetId.BINANCE_PEG_USDT,
            }
        ),
        BSC_TESTNET_CHAIN_ID: MappingProxyType(
            {
                "U": AssetId.TEST_U,
                "USDC": AssetId.TEST_USDC,
                "USDT": AssetId.TEST_USDT,
            }
        ),
    }
)

_DECIMAL_AMOUNT = re.compile(r"^[0-9]+(?:\.[0-9]+)?$")


def parse_asset_id(value: AssetId | str) -> AssetId:
    """Parse a canonical ID without applying network-specific symbol aliases."""
    if isinstance(value, AssetId):
        return value
    try:
        return AssetId(value)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"unknown canonical AssetId: {value!r}") from exc


def resolve_asset_alias(chain_id: int, alias: AssetAlias | str) -> AssetId:
    """Resolve a UI alias after the network context is known."""
    if chain_id not in _ALIASES:
        raise KeyError(f"no asset catalog registered for chain_id={chain_id}")
    try:
        return _ALIASES[chain_id][alias]
    except KeyError as exc:
        raise KeyError(f"unknown asset alias {alias!r} for chain_id={chain_id}") from exc


def get_asset(chain_id: int, asset_id: AssetId | str) -> PaymentAsset:
    """Return an active asset for a strict ``(chain_id, canonical AssetId)`` key."""
    return ASSET_CATALOG.get(chain_id, asset_id)


def get_asset_metadata(chain_id: int, asset_id: AssetId | str) -> PaymentAsset:
    """Return validated metadata, including diagnostic placeholder entries."""
    return ASSET_CATALOG.get_metadata(chain_id, asset_id)


def get_asset_by_address(chain_id: int, address: str) -> PaymentAsset:
    """Reverse-resolve an active token address; input casing is ignored."""
    return ASSET_CATALOG.by_address(chain_id, address)


def get_b402_asset(chain_id: int, asset_id: AssetId | str) -> B402PaymentAsset:
    """Return active metadata projected to the B402 execution contract."""
    return ASSET_CATALOG.get_b402(chain_id, asset_id)


def get_b402_asset_by_address(chain_id: int, address: str) -> B402PaymentAsset:
    """Reverse-resolve an active B402 token address; input casing is ignored."""
    return ASSET_CATALOG.by_b402_address(chain_id, address)


def list_assets(chain_id: int) -> tuple[PaymentAsset, ...]:
    """Return the immutable active catalog snapshot for one known network."""
    return ASSET_CATALOG.list(chain_id)


def known_eip3009_payment_tokens() -> frozenset[tuple[int, str]]:
    """Active catalog EIP-3009 ``(chain_id, checksum_address)`` domains."""
    return frozenset(
        (chain_id, get_b402_asset(chain_id, asset.asset_id).address)
        for chain_id in (BSC_MAINNET_CHAIN_ID, BSC_TESTNET_CHAIN_ID)
        for asset in list_assets(chain_id)
        if asset.eip3009_domain is not None and "eip3009" in asset.b402_methods
    )


def to_asset_atomic(chain_id: int, asset_id: AssetId | str, amount: str) -> int:
    """Convert a plain non-negative decimal string to exact atomic units."""
    if not isinstance(amount, str):
        raise TypeError("asset amount must be a decimal string")
    if _DECIMAL_AMOUNT.fullmatch(amount) is None:
        raise ValueError(f"invalid decimal amount: {amount!r}")

    asset = get_asset(chain_id, asset_id)
    whole, separator, fraction = amount.partition(".")
    if separator and len(fraction) > asset.decimals:
        raise ValueError(f"asset amount exceeds {asset.decimals} decimal places: {amount!r}")
    return int(whole) * 10**asset.decimals + int(fraction.ljust(asset.decimals, "0") or "0")
